jsapp/js/components/library/assetPublicButton.tsx

```diff
@@ -51,8 +56,20 @@ export default class AssetPublicButton extends React.Component<
     if (this.props.asset.uid === assetUid) {
       this.setState({
         isPublicPending: false,
+        // Public state of asset changed, now we await fresh permissions
         isAwaitingFreshPermissions: true,
       });
+
+      // We need to get fresh asset here, so that new permissions would be
+      // available for the button code. We rely on the fact that new asset
+      // would be passed through `props` and `componentWillReceiveProps` will
+      // unlock the button again.
+      //
+      // TODO: this flow should be improved, but it might require some more
+      // thought, as the asset data flow in the whole app should be redone
+      // (after very thorough planning). Unfortunately many places in the app
+      // have this problem.
```

Comment:
Agreed, and there are some problems with state getting out of sync.

- It's a little too easy to trigger an infinite loading spinner just by clicking around (breadcrumbs, asset in collection, back button)
- If you create a new empty collection, add an item to it, and go back to collection, you'll see the item in it, but if you try to make it public, you'll get an error because some part of the js still thinks the collection is empty.

Hitting "refresh" usually clears these out, but it would be nice to fix these.